### PR TITLE
NEW Form sudo mode

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-admin/issues/1898

### CODE REVIEW

File: src/Forms/GridField/GridField.php [hunk 1]
```diff
@@ -630,6 +631,21 @@ public function FieldHolder($properties = [])
             }
         }
 
+        // Set to read-only if sudo mode is required for the DataObject being managed
+        if (is_a($list, DataList::class)) {
```
Comment (emteknetnz) [inline src/Forms/GridField/GridField.php]: Scenario is a has_many relation GridField e.g. Page has_many Member, this is the top level gridfield, not the edit form

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php [hunk 2]
```diff
@@ -322,6 +322,11 @@ public function ItemEditForm()
         if ($cb) {
             $cb($form, $this);
         }
+
+        if (is_a($this->record, DataObject::class) && $this->record->getRequireSudoMode()) {
```
Comment (emteknetnz) [inline src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:330]: Scenario is the edit form in a gridfield i.e. editing an individual record

File: src/Forms/DefaultFormFactory.php [hunk 3]
```diff
@@ -47,6 +47,15 @@ public function getForm(RequestHandler $controller = null, $name = FormFactory::
         $validator = $this->getFormValidator($controller, $name, $context);
         $form = Form::create($controller, $name, $fields, $actions, $validator);
 
+        // Sudo mode
+        if (is_a($context['Record'], DataObject::class)) {
```
Comment (emteknetnz) [inline src/Forms/DefaultFormFactory.php:51]: Scenario is ... ?

Possibly should get rid of this as it may impact frontend forms

File: src/Security/MemberAuthenticator/SessionAuthenticationHandler.php [hunk 4]
```diff
@@ -82,11 +82,6 @@ public function logIn(Member $member, $persistent = false, HTTPRequest $request
         if (Member::config()->get('login_marker_cookie')) {
             Cookie::set(Member::config()->get('login_marker_cookie'), 1, 0);
         }
-
-        // Activate sudo mode on login so the user doesn't have to reauthenticate for sudo
-        // actions until the sudo mode timeout expires
-        $service = Injector::inst()->get(SudoModeServiceInterface::class);
```
Comment (emteknetnz) [inline src/Security/MemberAuthenticator/SessionAuthenticationHandler.php:88]: Agreed offline that sudo mode should NOT be automatically activated when you log in

File: src/Forms/SudoModePasswordField.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    protected $schemaComponent = 'SudoModePasswordField';
+
+    protected $extraClasses = [
+        'SudoModePasswordField',
+    ];
+
+    /**
+     * Whether the field should display with padding around it
+     */
+    private bool $panelPadded = true;
```
Comment (emteknetnz) [inline src/Forms/SudoModePasswordField.php]: The "panelPadded" functionality in [this PR](https://github.com/silverstripe/silverstripe-cms/pull/3053) is related to making the field looking pretty "inline" which is only used in that PR which I think we should close. If we close that, then the panelPadded functionality should be removed from this PR.

File: src/Forms/Form.php [hunk 6]
```diff
@@ -317,6 +320,51 @@ public function __construct(
...
+     * unless the current user previously activated sudo mode.
+     */
+    public function requireSudoMode(): void
+    {
+        // Check that the current request handler for the form is one that's used
+        // in an admin context where sudo mode makes sense
+        $classes = [
+            LeftAndMain::class,
+            GridFieldDetailForm_ItemRequest::class,
+        ];
```
Comment (GuySartorelli) [inline src/Forms/Form.php:339]: I don't think this distinction should be made - instead, provide a `setCanUseSudoMode()` and `getCanUseSudoMode()` - true by default, but can be turned off on a per-form basis.

i.e. we should be providing these protections even on the front-end if people are exposing member edit forms on the front-end - and people can choose to _explicitly disable that_ on the front-end if they want to, in which case they're responsible for their own security.

File: src/Forms/Form.php [hunk 7]
```diff
@@ -317,6 +320,51 @@ public function __construct(
...
+        $enableSudoMode = false;
+        foreach ($classes as $class) {
+            if (class_exists($class) && is_a($this->getController(), $class)) {
```
Comment (GuySartorelli) [inline src/Forms/Form.php]:
```suggestion
            if (is_a($this->getController(), $class)) {
```
No need to check if the class exists - if it doesn't exist, `is_a` will return false, because the controller isn't that class.

File: src/Forms/Form.php [hunk 8]
```diff
@@ -317,6 +320,51 @@ public function __construct(
...
+        if ($service->check($session)) {
+            return;
+        }
+        // If sudo mode is not active, make the form readonly and add a sudo mode password field
+        $this->makeReadonly();
+        $name = SudoModePasswordField::FIELD_NAME;
+        $fields = $this->Fields();
+        if ($fields->fieldByName($name)) {
+            $fields->removeByName($name);
+        }
```
Comment (GuySartorelli) [inline src/Forms/Form.php]: Why are we removing the field? We should use a field name that is extremely unlikely to clash with existing form field names in the first place.

File: src/Forms/Form.php [hunk 9]
```diff
@@ -317,6 +320,51 @@ public function __construct(
...
+        $field = SudoModePasswordField::create($name);
+        $field->setForm($this);
+        $first = $fields->first();
+        if ($first) {
+            $fields->insertBefore($first->getName(), $field);
+        } else {
+            $fields->push($field);
+        }
```
Comment (GuySartorelli) [inline src/Forms/Form.php]:
```suggestion
        $fields->unshift($field);
```
We shouldn't need to explicitly set form, I think? If that is needed here for some reason, we probably need an inline code comment explaining why because that's highly unusual.

Also, we can just `unshift` directly to the top of the fieldlist.

File: src/Forms/GridField/GridField.php (hunk 1 above)
Comment (GuySartorelli) [reply src/Forms/GridField/GridField.php]: We shouldn't be checking based on the list class - e.g. someone might go
```php
$memberList = Member::get()->filterByCallback(/*...*/);
```
and then pass that `ArrayList` into the gridfield.

File: src/Forms/GridField/GridField.php [hunk 10]
```diff
@@ -634,6 +635,23 @@ public function FieldHolder($properties = [])
             }
         }
 
+        
+        // Set GridField to read-only if sudo mode is required for the DataObject being managed
+        // and sudo mode is not active
+        if (is_a($list, DataList::class)) {
+            /** @var DataObject $obj */
+            $obj = Injector::inst()->create($list->dataClass);
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridField.php]:
```suggestion
        $modelClass = $this->getModelClass();
        if (is_a($modelClass, DataObject::class, true)) {
            /** @var DataObject $obj */
            $obj = Injector::inst()->create($modelClass);
```

File: src/ORM/DataObject.php [hunk 11]
```diff
@@ -1401,6 +1408,14 @@ public function populateDefaults()
         return $this;
     }
 
+    /**
+     * Whether the DataObject requires sudo mode to be active in order to be modified in admin
+     */
+    public function getRequireSudoMode(): bool
+    {
+        return $this->config()->get('require_sudo_mode');
+    }
```
Comment (GuySartorelli) [inline src/ORM/DataObject.php]: This and the config should be on `ViewableData` instead. e.g. you might want sudo mode on a model that interacts with a third-party API.

File: src/Forms/GridField/GridField.php [hunk 12]
```diff
@@ -634,6 +635,23 @@ public function FieldHolder($properties = [])
...
+            if ($obj->getRequireSudoMode()) {
+                $session = Controller::curr()?->getRequest()?->getSession();
+                if ($session) {
+                    $service = Injector::inst()->get(SudoModeServiceInterface::class);
+                    if (!$service->check($session)) {
+                        $this->performReadonlyTransformation();
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridField.php]: This method changes the components that are available - so this should probably be the very first thing that happens inside `FieldHolder()`.

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php (hunk 2 above)
Comment (GuySartorelli) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:330]: This should be after `updateItemEditForm` because that extension hook might add additional fields, which we want to be included in the readonly transformation.

File: src/Forms/SudoModePasswordField.php [hunk 13]
```diff
@@ -0,0 +1,25 @@
+<?php
+
+namespace SilverStripe\Forms;
+
+use SilverStripe\Forms\FormField;
+
+class SudoModePasswordField extends FormField
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php]: Why not extend `PasswordField`?

File: src/Forms/SudoModePasswordField.php [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+{
+    public const FIELD_NAME = 'SudoModePasswordField';
+    
+    protected $schemaComponent = 'SudoModePasswordField';
+
+    protected $extraClasses = [
+        'SudoModePasswordField',
+    ];
+
+    public function __construct($name, $title = null, $value = '')
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php]: Why accept any args if we're just gonna ignore them?

File: src/Forms/SudoModePasswordField.php [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+
+class SudoModePasswordField extends FormField
+{
+    public const FIELD_NAME = 'SudoModePasswordField';
+    
+    protected $schemaComponent = 'SudoModePasswordField';
+
+    protected $extraClasses = [
+        'SudoModePasswordField',
+    ];
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php]: The existing pattern is to call `$this->addExtraClass()` in the constructor instead.
```suggestion
```

File: src/Forms/SudoModePasswordField.php [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+    public function __construct($name, $title = null, $value = '')
+    {
+        // Name must be "SudoModePasswordField" as there's logic elsewhere expecting this
+        // $name and $value are set to null as the react component does not use these arguments
+        parent::__construct(SudoModePasswordField::FIELD_NAME, null, null);
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php]:
```suggestion
        parent::__construct(SudoModePasswordField::FIELD_NAME);
```
No need to be explicitly about passing `null` - those are already the default args.

File: src/Forms/SudoModePasswordField.php [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    protected $schemaComponent = 'SudoModePasswordField';
+
+    protected $extraClasses = [
+        'SudoModePasswordField',
+    ];
+
+    public function __construct($name, $title = null, $value = '')
+    {
+        // Name must be "SudoModePasswordField" as there's logic elsewhere expecting this
+        // $name and $value are set to null as the react component does not use these arguments
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php]:
```suggestion
        // Name must be "SudoModePasswordField" as there's logic elsewhere expecting this.
        // $title and $value are set to null as the react component does not use these arguments.
```

File: src/Forms/SudoModePasswordField.php [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+        parent::__construct(SudoModePasswordField::FIELD_NAME, null, null);
+        // Set title to empty string to avoid rendering a label before the react component has loaded
+        $this->setTitle('');
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php:19]: That feels like the wrong way to change how a thing renders. I assume the label is being rendered in a template? So we should probably just make sure the template is correct for this new class instead.

File: src/Security/SudoMode/SudoModeService.php [hunk 19]
```diff
@@ -45,6 +49,11 @@ public function activate(Session $session): bool
         return true;
     }
 
+    public function deactivate(Session $session): void
+    {
+        $session->set(SudoModeService::SUDO_MODE_SESSION_KEY, null);
+    }
```
Comment (GuySartorelli) [inline src/Security/SudoMode/SudoModeService.php:55]: What is this new method for? After it's activated, shouldn't it remain active for the full request cycle?

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/Form.php (hunk 6 above)
Comment (emteknetnz) [reply src/Forms/Form.php:339]: Adding this to the front-end would be an increase on the original scope. While it's not explicitly listed in the ACs, the intention of this work was exclusively for inside the CMS.

It's also something that I wouldt even want to do, even in a major, at least not as an opt-out, because at that point we're making a functional change to someones frontend website, not just the CMS.

If there was demand for a sudo-mode for member data managed on the frontend, I think that should be done as a completely separate issue

File: src/Forms/Form.php (hunk 8 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: We already are using a pretty unique field name. However unlikely is not the same as impossible. Also there may be some weird exploit that involves fields with the same field name. So doing this to be sure

Comment (GuySartorelli) [reply src/Forms/Form.php]: We shouldn't be removing fields that are intentionally there. An exception is probably better if you're that worried about collisions - and the exception will happen automatically by adding the duplicate-named field.

File: src/Forms/Form.php (hunk 9 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: Updated to use unshift()

setForm() is not unusual at all e.g. FieldList $fields->setForm($this) is called in Form::__construct() which will loop $fields and call $field->setForm($this);

I don't see why it needs a comment fields normally get attached for their parent form

Comment (GuySartorelli) [reply src/Forms/Form.php]: It's unusual to _manually_ call `setForm()` when adding a field to a fieldlist. I don't think I've ever seen that - you usually just add the field to the fieldlist and let the form/fieldlist/field relations sort themselves out

File: src/ORM/DataObject.php (hunk 11 above)
Comment (emteknetnz) [reply src/ORM/DataObject.php]: Maybe, though I feel like we're going outside the original intended scope by considering scenarios like this. Third party gridfield managed data is likely a very niche use case at this point, so I'm not particularly keen to increase the scope of this work. Remember the original intention of this was primarily around mitigating privilege escalation attacks, e.g. prevent new malicious admins from being created, not around general data protection.

File: src/Forms/SudoModePasswordField.php (hunk 13 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: I cannot remember the exact reason. It did extended PasswordField at first, though something about that wasn't great so I changed it to FormField.

File: src/Forms/SudoModePasswordField.php (hunk 14 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: I prefer to not change method signatures of subclasses

File: src/Security/SudoMode/SudoModeService.php (hunk 19 above)
Comment (emteknetnz) [reply src/Security/SudoMode/SudoModeService.php:55]: When logging in with a fresh member for the first time (possibly other scenarios too), if the MFA module is installed you'll be prompted to sign up for MFA on the frontend. If the user does so, then MFA will be silently activated so that the sudo checks in controller endpoints pass.  

This still needs to happen, though as we're removing the 'automatically activate sudo mode when logging in' bit, we need to deactivate the silently activated sudo mode.

File: src/Forms/Form.php (hunk 7 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: Updated

File: src/Forms/Form.php (hunk 8 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: Updated

File: src/Forms/GridField/GridField.php (hunk 12 above)
Comment (emteknetnz) [reply src/Forms/GridField/GridField.php]: Updated

File: src/Forms/SudoModePasswordField.php (hunk 15 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Updated

File: src/Forms/SudoModePasswordField.php (hunk 14 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Updated, will remove all the args

File: src/Forms/SudoModePasswordField.php (hunk 17 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Updated

File: src/Forms/SudoModePasswordField.php (hunk 16 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Updated

File: src/Forms/SudoModePasswordField.php (hunk 18 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php:19]: Trying to use a custom SudoModePasswordField_holder.ss, or using `function FieldHolder() ... return $this->Field()) ends up breaking a bunch of stuff.  It's much simpler and less code to just use the use a single line of code to achieve the same outcome

File: src/Forms/GridField/GridFieldDetailForm_ItemRequest.php (hunk 2 above)
Comment (emteknetnz) [reply src/Forms/GridField/GridFieldDetailForm_ItemRequest.php:330]: Updated

File: src/Forms/GridField/GridField.php [hunk 20]
```diff
@@ -521,6 +523,28 @@ public function FieldHolder($properties = [])
...
         $this->extend('onBeforeRenderHolder', $this, $properties);
 
+        // Set GridField to read-only if sudo mode is required for the DataObject being managed
+        // and sudo mode is not active
+        $modelClass = null;
+        try {
+            $modelClass = $this->getModelClass();
+        } catch (LogicException) {
+            // noop
+        }
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridField.php]: What scenario did you have where a model class wasn't defined by this stage? That shouldn't happen unless something is misconfigured.

File: src/Forms/SudoModePasswordField.php [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+
+use SilverStripe\Forms\FormField;
+
+class SudoModePasswordField extends FormField
+{
+    public const FIELD_NAME = 'SudoModePasswordField';
+    
+    protected $schemaComponent = 'SudoModePasswordField';
+
+    public function __construct($name, $title = null, $value = '')
```
Comment (GuySartorelli) [inline src/Forms/SudoModePasswordField.php]:
```suggestion
    public function __construct()
```
As per https://github.com/silverstripe/silverstripe-framework/pull/11582#discussion_r1957440288

File: src/Forms/SudoModePasswordField.php (hunk 17 above)
Comment (GuySartorelli) [reply src/Forms/SudoModePasswordField.php]: Looks like you forgot to push

File: src/Security/SudoMode/SudoModeService.php (hunk 19 above)
Comment (GuySartorelli) [reply src/Security/SudoMode/SudoModeService.php:55]: MFA being activated shouldn't disable/activate/bypass sudo mode, should it?

> This still needs to happen

Why? Sounds a bit pointless if we're immediately deactivating it anyway?

If there's something in MFA that requires MFA to be active to set it up, sounds like the MFA module should be updated so that doesn't happen if being set from that front-end prompt.

File: src/Forms/SudoModePasswordField.php (hunk 13 above)
Comment (GuySartorelli) [reply src/Forms/SudoModePasswordField.php]: Can you please try it again and take a note of what goes wrong? "Something about that wasn't great" doesn't help me review this - I need to know why it doesn't work so I can determine if there's another way around the problem you were facing, and if this is the correct approach or not.

File: src/ORM/DataObject.php (hunk 11 above)
Comment (GuySartorelli) [reply src/ORM/DataObject.php]: The scope of the work doesn't really change, does it? You just move some code from being on `DataObject` to instead be on `ViewableData` and some `is_a(... DataObject::class)` checks to be `is_a(... ViewableData::class)` checks.

> Remember the original intention of this was primarily around mitigating privilege escalation attacks

I think that allowing developers to protect their own models from privilege escalation attacks fits this intention perfectly.

> ... i.e. prevent new malicious admins from being created, not around general data protection.

If you're saying the ability to set this config on models other than the ones we specifically control is out of scope, then the whole solution is wrong. I think that's probably not what you're intending to say but that's definitely how this reads to me.

File: src/Forms/Form.php (hunk 6 above)
Comment (GuySartorelli) [reply src/Forms/Form.php:339]: If we're excluding front-end usage from the scope, we need to add some language to the docs to the effect of "this functionality is intended for use in the CMS - using it in the frontend is not currently supported"

That said, it feels like the effort required to allow usage in the front-end is very minimal.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/Form.php (hunk 6 above)
Comment (emteknetnz) [reply src/Forms/Form.php:339]: Have added a `> [!NOTE]` to the 04_Sudo_Mode.md docs. Left the changelog unaltered as that already has plenty of mentions of 'the CMS'

File: src/Forms/GridField/GridField.php (hunk 20 above)
Comment (emteknetnz) [reply src/Forms/GridField/GridField.php]: 1) SilverStripe\Forms\Tests\GridField\GridFieldTest::testGridFieldCustomFragments
LogicException: GridField doesn't have a modelClassName, so it doesn't know the columns of this grid.

https://github.com/creative-commoners/recipe-kitchen-sink/actions/runs/13359908314/job/37307887025#step:12:211

$this->getModelClass() can throw an exception so we need to handle it

File: src/Forms/SudoModePasswordField.php (hunk 13 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Seems fine, have updated to extend PasswordField.  Have added support for passing through the 'autocomplete' config on PasswordField to the SudoPasswordField react component

File: src/Forms/SudoModePasswordField.php (hunk 21 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Done

File: src/Forms/SudoModePasswordField.php (hunk 17 above)
Comment (emteknetnz) [reply src/Forms/SudoModePasswordField.php]: Updated

File: src/ORM/DataObject.php (hunk 11 above)
Comment (emteknetnz) [reply src/ORM/DataObject.php]: I just think that it's too confusing for the average user. 99.9% of the time we're going to be setting this on a DataObject, and developers understand how DataObjects work

Trying to explain that "You can manage the not-at-all-confusingly-named-ViewableData in a GridField which could be third party API data, which you probably had no idea could be GridField managed" and they won't have a clue what you're talking about.

We're already introducing a new concept (and basically a new UX annoyance) of extra protected data via the changelog. I don't want to have to also explain how ViewableData is actually sort of this abstraction around data which may not actually be viewable. We're still in CMS 5 remember.

> If you're saying the ability to set this config on models other than the ones we specifically control is out of scope

The fact you can do that is essentially just a nice bonus feature due the implementation using config on DataObject as opposed to on a Form or a Controller.  It's not the original goal of this work which was to make users have to re-enter their password before modifying member/group etc data

File: src/Security/SudoMode/SudoModeService.php (hunk 19 above)
Comment (emteknetnz) [reply src/Security/SudoMode/SudoModeService.php:55]: In the MFA flow there are controller endpoints to see if sudo mode is active or not. If it's not active then you cannot register.

Remember that MFA can be used inside and outside of the CMS.  The activate/deactive bit is for outside of the CMS when logging in for the first time and registering mfa

File: src/ORM/DataObject.php (hunk 11 above)
Comment (emteknetnz) [reply src/ORM/DataObject.php]: Actually have reconsidered and moved config + method to ViewableData, with the caveat that we should keep our docs as is and keep mentioning DataObject rather than ViewableData for the reasons mentioned above as not confusing people is my main concern

Comment (GuySartorelli) [reply src/ORM/DataObject.php]: > You can manage the not-at-all-confusingly-named-ViewableData in a GridField which could be third party API data, which you probably had no idea could be GridField managed

... yes, it will be confusing if you're making it intentionally confusing to try prove your point 😅

How about "You can configure any `ViewableData`, including your `DataObject` classes..."

We'll pick up the documentation concerns in the docs PR though, no need to discuss that here.

File: src/View/ViewableData.php [hunk 22]
```diff
@@ -261,6 +268,14 @@ public function hasDynamicData(string $field): bool
         return array_key_exists($field, $this->dynamicData);
     }
 
+    /**
+     * Whether the model requires sudo mode to be active in order to be modified in admin
+     */
+    public function getRequireSudoMode(): bool
+    {
+        return $this->config()->get('require_sudo_mode');
+    }
```
Comment (GuySartorelli) [inline src/View/ViewableData.php]: Why do we have a method that just returns a config value? Shouldn't we just check the config directly instead? Or is the idea that people can override this method to do some complex conditional logic in here if they want to?

File: src/Forms/Form.php [hunk 23]
```diff
@@ -317,6 +320,41 @@ public function __construct(
         $this->setupDefaultClasses();
     }
 
+    /**
+     * Make the form require sudo mode, which will make the form readonly and add a sudo mode password field
+     * unless the current user previously activated sudo mode.
```
Comment (GuySartorelli) [inline src/Forms/Form.php:327]:
```suggestion
     * unless the current user previously activated sudo mode.
     *
     * Note that if the parent request handler for this form isn't LeftAndMain or GridFieldDetailForm_ItemRequest,
     * sudo mode will not be required by this form.
```

File: src/Forms/GridField/GridField.php (hunk 20 above)
Comment (GuySartorelli) [reply src/Forms/GridField/GridField.php]: > $this->getModelClass() can throw an exception so we need to handle it

We throw a lot of exceptions explicitly to _not_ be caught by our code, so that developers can be aware of misconfigurations or logic errors in their code. So just because we're throwing an exception doesn't mean we need to catch it.

That said, it looks like that test is (inadvertently) showing that GridField can be used in a way that doesn't require showing columns from the objects held in its list - which would be an _extreme_ edge case but I guess there's no reason to break that flexibility here.

I'll add a suggestion to update the `noop` comment so I don't see this in the future and freak out lol

File: src/Forms/GridField/GridField.php [hunk 24]
```diff
@@ -521,6 +523,28 @@ public function FieldHolder($properties = [])
     {
         $this->extend('onBeforeRenderHolder', $this, $properties);
 
+        // Set GridField to read-only if sudo mode is required for the DataObject being managed
+        // and sudo mode is not active
+        $modelClass = null;
+        try {
+            $modelClass = $this->getModelClass();
+        } catch (LogicException) {
+            // noop
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridField.php]:
```suggestion
            // noop - it's possible to have a gridfield with custom components that don't rely on columns from the records in the list.
```

File: src/Forms/Form.php (hunk 9 above)
Comment (GuySartorelli) [reply src/Forms/Form.php]: This still needs to be addressed. If setting the form here is necessary for some reason, please add a code comment explaining why, as I believe that should be handled automatically by adding the field to a fieldlist which belongs to a  form.

File: src/Security/SudoMode/SudoModeService.php (hunk 19 above)
Comment (GuySartorelli) [reply src/Security/SudoMode/SudoModeService.php:55]: This discussion will continue in https://github.com/silverstripe/silverstripe-mfa/pull/587#discussion_r1955289016 which is where the MFA changes need to happen

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/Form.php (hunk 23 above)
Comment (emteknetnz) [reply src/Forms/Form.php:327]: Updated

File: src/Forms/GridField/GridField.php (hunk 24 above)
Comment (emteknetnz) [reply src/Forms/GridField/GridField.php]: Updated

File: src/View/ViewableData.php (hunk 22 above)
Comment (emteknetnz) [reply src/View/ViewableData.php]: LeftAndMain has `if (is_a($record, DataObject::class) && $record->getRequireSudoMode()) {` - I think it's nicer to call a method in that context rather than reading another classes config

I have just made an update to change this to use static::config() instead

**Comment by GuySartorelli** [comment]
https://github.com/silverstripe/silverstripe-framework/pull/11582#discussion_r1958817736 still not addressed

File: src/Forms/Form.php (hunk 9 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: Have added a code comment

File: src/Forms/Form.php [hunk 25]
```diff
@@ -317,6 +322,58 @@ public function __construct(
...
+        // Check if sudo mode is currently enabled
+        $service = Injector::inst()->get(SudoModeServiceInterface::class);
+        $session = $this->getRequest()->getSession();
+        if ($service->check($session)) {
+            return;
+        }
+        // If sudo mode is not active, make the form readonly and add a sudo mode password field
+        // Modify the GridFieldViewButton on any GridFields so that it doesn't suffix the view URL with 'view'
+        // This allows us to gracefully reload a form in readonly mode when sudo mode is activated
+        foreach ($this->Fields()->flattenFields() as $field) {
```
Comment (GuySartorelli) [inline src/Forms/Form.php]: Is there a way we can do this without calling `flattenFields()`?
That method results in creating a new field list and adding the fields to it, which results in calling `setContainerFieldList()` on the fields, setting the wrong fieldlist for them.

E.g. could use `recursiveWalk()` probably.

File: src/Forms/Form.php [hunk 26]
```diff
@@ -317,6 +322,58 @@ public function __construct(
...
+            if (!is_a($field, GridField::class)) {
+                continue;
+            }
+            $config = $field->getConfig();
+            $config->removeComponentsByType(GridFieldViewButton::class);
+            $viewButton = GridFieldViewButton::create();
+            $viewButton->setSuffixViewToUrl(false);
+            $config->addComponent($viewButton);
+        }
```
Comment (GuySartorelli) [inline src/Forms/Form.php]: This is fine for if the _whole form_ is readonly, but doesn't work for this scenario:

```php
    public function getCMSFields()
    {
        $fields = parent::getCMSFields();
        $fields->addFieldToTab('Root.Main', GridField::create('Members', 'Members', Member::get(), GridFieldConfig_RecordEditor::create()));
        return $fields;
    }
```

That gridfield is still readonly with a normal unaltered view button component.

I think this code actually doesn't need to be in `Form` at all - just in `GridField::Fieldholder()`

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/Form.php (hunk 25 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: Moving code to `GridField::Fieldholder()`

File: src/Forms/Form.php (hunk 26 above)
Comment (emteknetnz) [reply src/Forms/Form.php]: Yeah much nicer in `GridField::Fieldholder()`, will do there

File: src/Forms/Form.php (hunk 25 above)
Comment (GuySartorelli) [reply src/Forms/Form.php]: This is still here?

File: src/Forms/Form.php (hunk 26 above)
Comment (GuySartorelli) [reply src/Forms/Form.php]: Looks like you forgot to push

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Forms/GridField/GridField.php [hunk 27]
```diff
@@ -552,6 +580,13 @@ public function FieldHolder($properties = [])
             if ($item instanceof GridFieldPaginator) {
                 $total = $item->getTotalItems();
             }
+            if ($sudoModeTransformation) {
+                // Modify the GridFieldViewButton on any GridFields so that it doesn't suffix the view URL with 'view'
+                // This allows us to gracefully reload a form in readonly mode when sudo mode is activated
+                if ($item instanceof GridFieldViewButton) {
```
Comment (GuySartorelli) [inline src/Forms/GridField/GridField.php:586]: I'd normally ask for this to be combined into a single conditional statement but I think we're both tired of this review cycle, and the change is only a very minor code style change, so will leave as is.

**Review by GuySartorelli — APPROVED** [review]
LGTM, will give one last look locally.
